### DISCUSSION TITLE
How to change the velocity of an entity

### DISCUSSION THREAD

**Discussion by asd988** [FabricMC/fabric-api#1949] [Mod Dev Support] [answered]
I can't manage to figure it out, `addVelocity` and `setVelocity` methods doesn't work for some reason. (I have this code in an items `usageTick` method) I have tried so many ways to make it work but it failed. My guess is that the server isn't synced with this change, because after I print out the velocity after it has been set it will be what I set it, while in game nothing changes.

**Comment by warjort** [comment]
> My guess is that the server isn't synced with this change

Are you changing it on the server?

**Reply by asd988** [reply]
It runs on render thread l, so I think not.

**Comment by warjort** [comment]
The server is what holds the real data and does stuff.
The client's data is just a copy so it can draw things.

**Reply by asd988** [reply]
Alright, so how am I meant to do it then? I tried a dozen different ways and either it didn't work or either it was incredibly buggy and inconsistent.
This is the part of the code where I want to set the velocity:
```java
List<Entity> affectedEntities = world.getOtherEntities(user, new Box(
        user.getX()-5, user.getY()-4, user.getZ()-5,
        user.getX()+5, user.getY()+6, user.getZ()+5
));
for (Entity entity : affectedEntities) {
    entity.setVelocity(new Vec3d(0d, 1d ,0d));
    entity.velocityModified = true;
}
```

**Comment by warjort** [comment]
This code snippet doesn't explain where it is being called/used.
And its hard to say what is wrong if you only describe it as "incredibly buggy".

If you want to modify game state from the client then you need to send a custom network message to the server.
I assume you are not using one of the existing actions like clicking on or attacking an entity that will do this for you.

As to the details, have a look at something like ````LivingEntity.takeKnockBack()```` that modifies an entity's velocity.

One thing I notice in your code is that you setVelocity() which does not modify Entity.velocityDirty while addVelocity() does.
The knockback code mentioned above also uses setVelocity() but sets the dirty flag.
I think this just determines whether updates to the client happen immediately or are delayed according to their tracking config on the entity type?

**Reply by asd988** [reply]
> If you want to modify game state from the client then you need to send a custom network message to the server.

How do I do that? 

> As to the details, have a look at something like LivingEntity.takeKnockBack() that modifies an entity's velocity.

All I can see there, is that for some reason it works and here it doesn't, apart from `velocityDirty`, but I already tried that and it obviously didn't work.


 >  And its hard to say what is wrong if you only describe it as "incredibly buggy".

Ignore that, I was talking about my not working "solutions" to the problem.

Also, I think we should ignore that it runs in `usageTick` because I literally injected into `tick` method of `MinecraftServer` (and other tick methods of other classes), and it still didn't change anything.

**Reply by warjort** [reply]
For networking see the wiki:
https://fabricmc.net/wiki/tutorial:networking

Vanilla uses the PlayerInteractEntityC2SPacket to send entity interaction events to the server.

**Reply by asd988** [reply]
Thanks!

But, do you know why doesn't this work? (I injected this into the head of `tick` method inside `MinecraftServer` class)
This works on player, but if I put enemies inside it they won't budge. (this runs on server thread)
```java
if (MinecraftClient.getInstance().world != null) {
    List<Entity> affectedEntities = MinecraftClient.getInstance().world.getOtherEntities(null, new Box(
            0, 0, 0,
            5, 15, 5
    ));
    for (Entity entity : affectedEntities) {
        entity.velocityDirty = true;
        entity.setVelocity(new Vec3d(0d, 1d ,0d));
    }
}
```

**Reply by warjort** [reply]
> MinecraftClient.getInstance()

Not the server

**Reply by asd988** [reply]
Oh wow... this comment surprisingly helped a lot!

**Comment by warjort** [comment]
> I have this code in an items usageTick method

I see you mentioned earlier where the code is used.
The Item.usageTick() isn't something I am familiar with.
Looking at the MC 1.18 code, it is only used by the Crossbow with the method first checking ````!world.isClient()````
From that I don't know if it usable on the client? Mojang don't use it that way.

**Comment by asd988** [comment, marked as answer]
Thanks SO MUCH to @warjort. The solution is actually very obvious, I don't know how did I miss this...

I got the entities from `ClientWorld` (meaning I could only read their data not write it), while I should've got them from `ServerWorld` (where I can actually write it).
